### PR TITLE
UME-2 interaction module

### PR DESCRIPTION
**Pull request by karinazad**
## Description
Add UME Interaction Module that takes in pre-trained UME sequence encoders (and later sequence-structure encoders) and adds cross-attention. Training objective is MLM on individual sequences. This implementation does not assume a shared tokenizer. 

in pseudocode:
```py
# input
input: masked sequence1, masked sequence2, modality1, modality2

# encode individually
x1 = encoder(seq1), x2 = encoder2(seq2)

# cross attention update
x1, x2 = cross-attention-update(x1,x2)

# compute loss individually
logits1 = decoder1(x1), logits2=decoder2(x2)
loss = loss1 + loss 2 / 2
```

## Type of Change
- [ ] Bug fix
- [x] New feature  
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

### CODE REVIEW

File: src/lobster/callbacks/_peer_evaluation_callback.py
Comment (karinazad) [inline src/lobster/callbacks/_peer_evaluation_callback.py:1]: @taylormjs can I pls get a review? I tried to update PEER to use `embed_sequences` so that it doesn't rely on UME tokenizer

File: src/lobster/model/ume2/ume_sequence_encoder.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        # Load state_dict
+        state_dict = map_checkpoint_keys(state_dict, original_prefix="encoder.neobert.", new_prefix="")
+        encoder.neobert.load_state_dict(state_dict)
+
+        return encoder
+
+    def forward(
+        self, input_ids: Tensor, attention_mask: Tensor, return_auxiliary_tasks: bool = False, **kwargs
+    ) -> Tensor:
+        output = self.neobert(input_ids=input_ids, attention_mask=attention_mask, **kwargs)
```
Comment (taylormjs) [inline src/lobster/model/ume2/ume_sequence_encoder.py:87]: I wonder if we should have an object for outputs.  So output = Output("encoder_output": neobert_output, "aux_output": aux_output). That way, you can access both just be calling UMESequenceEncoderModule(...) in case you just want the encoder outputs

**Review by taylormjs — COMMENTED** [review]
Looking good! I'd recommend adding more granular tests, such as checking the inputs and attention masks into each module, including attention blocks (like scaled_dot_product_attention). You could also integrate the model._pooler and post_train._unfreezing code

File: src/lobster/datasets/s3_datasets/atomica.py [hunk 2]
```diff
@@ -13,6 +14,7 @@
 class Atomica(UMEStreamingDataset):
     SEQUENCE_KEY_1 = "sequence1"
     SEQUENCE_KEY_2 = "sequence2"
+    SEQUENCE_KEY_3 = "sequence3"
```
[the remark this answers is not in the document]
Comment (karinazad) [reply src/lobster/datasets/s3_datasets/atomica.py:17]: they are pretty infrequent (I think less than 5%) and there can be more than 3 - sometimes up to 5

File: src/lobster/model/ume2/auxiliary_tasks.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        layers.append(nn.Linear(self.hidden_size, output_dim))
+
+        self.head = nn.Sequential(*layers)
+
+    def forward(self, x: Tensor) -> Tensor:
+        match self.pooling:
+            case "cls":
+                x = x[:, 0, :]
+            case "mean":
```
[the remark this answers is not in the document]
Comment (karinazad) [reply src/lobster/model/ume2/auxiliary_tasks.py:67]: good catch! and we can make a separate PR for replacing this with post_train modules

File: src/lobster/model/ume2/ume_sequence_encoder.py (hunk 1 above)
Comment (karinazad) [reply src/lobster/model/ume2/ume_sequence_encoder.py:87]: we could have an output format similar to what NeoBERT has but for aux tasks, it would have to be a dict inside it anyway probably since we have variable task names?

File: src/lobster/callbacks/_auxiliary_task_loss_weight_scheduler.py
[the remark this answers is not in the document]
Comment (karinazad) [reply src/lobster/callbacks/_auxiliary_task_loss_weight_scheduler.py:1]: oops

File: src/lobster/model/ume2/ume_interaction.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+class UMEInteraction(nn.Module):
+    SUPPORTED_MODALITIES = [Modality.AMINO_ACID, Modality.SMILES]
+
+    def __init__(
+        self,
+        checkpoints: dict[str | Modality, str] | None = None,
+        supported_modalities: list[str | Modality] | None = None,
+        encoder_kwargs: dict[str | Modality, dict] | None = None,
+        freeze_molecular_encoders: bool = False,
```
[the remark this answers is not in the document]
Comment (karinazad) [reply src/lobster/model/ume2/ume_interaction.py:21]: yep definitely! we can make a new PR for that

File: src/lobster/model/ume2/ume_sequence_encoder.py (hunk 1 above)
Comment (taylormjs) [reply src/lobster/model/ume2/ume_sequence_encoder.py:87]: yeah good point, cool we can address a standard Output in another PR, where on entry is just aux tasks

File: src/lobster/model/ume2/_cross_attention.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+            attn_weights = attn_scores.softmax(dim=-1)
+            attn = attn_weights @ xv.permute(0, 2, 1, 3)
+            attn = attn.transpose(1, 2)
+        else:
+            # Use SDPA
+            attn = scaled_dot_product_attention(
+                query=xq.transpose(1, 2),
+                key=xk.transpose(1, 2),
+                value=xv.transpose(1, 2),
+                attn_mask=attn_mask.bool(),
```
Comment (taylormjs) [inline src/lobster/model/ume2/_cross_attention.py:186]: @karinazad I'm not 100% sure on if this attn mask is correct or needs to be inverted. Flagging for now
